### DISCUSSION TITLE
PHP attributes Override >= 8.3 & Deprecated >=8.4 are the only php attributes that are usable without a non attributes alternative? Any other ideas?

### DISCUSSION THREAD

**Discussion by ghost** [laravel/framework#52193] [Q&A]
Since the introduction of attributes I always stated that they should be used only when they are the only option.

Finally I found one case when they are the only option (that I know of).

https://www.php.net/manual/en/class.override.php

Do you know other situations when the attributes would be the only option?

(side note: this attribute does not need reflection from what I see, it is embedded in php).

@henzeb said also: ArrayShape
@dmitrychurkin said: Deprecated

**Comment by henzeb** [comment]
ArrayShape. For intelligence. No reflection needed, as it is for IDE's

**Reply by ghost** [reply]
With the mention that this is not a native attribute of php but a class from
use JetBrains\PhpStorm\ArrayShape;

**Reply by henzeb** [reply]
Indeed, but it would be nice if other ide's can interpret those too.

**Comment by henzeb** [comment]
Got another use case. In my newly build package I have created an EncryptsData Attribute. The only moment it is used is when you rotate the encryption on your models where you have a mutator that handles encryption. You can also avoid it by creating a full cast object.

https://packagist.org/packages/henzeb/laravel-encrypted-data-rotator

I think its an exception to the rule. It is not affecting performance of your application when used in a command line or queue.

**Reply by ghost** [reply]
Nice. It is a case used only when you rotate by running that command. I am just curious. Is this the only solution to solve this problem?  I understand that it must run only when you run the command and not when the application is running.

**Reply by henzeb** [reply]
there is also a solution to add a parameter with a list of  mutators, but I find that a bit messy. the casts are also automatically detected.

I doubt there are problems that only have one solution that involves Attributes anyway.

**Reply by ghost** [reply]
I see it as an acceptable exception to use attributes on something (as "flags") that will be discovered only on certain manual initiated actions and not on each request because they will not add any overhead to the execution code in normal circumstances.

**Comment by ghost** [comment]
@henzeb a colleague found another use case. Open api generation of swagger documentation via attributes. It is under the same logic as is your example for one time command for encrypting data.

**Comment by ghost** [comment]
@henzeb bad news

![image](https://github.com/user-attachments/assets/32dc7c4d-555e-4592-8a18-2ce84fba22f9)

**Reply by henzeb** [reply]
Yeah I saw it. I wonder how much impact this would actually have even when not used.

**Reply by ghost** [reply]
If this will be for each instantiation of any class (even models).... then... no comment...

**Reply by ghost** [reply]
@henzeb this reminds me of: http://harmful.cat-v.org/software/c++/I_did_it_for_you_all

**Comment by dmitrychurkin** [comment]
@marius-mcp Starting from 8.4 Deprecated as well https://www.php.net/manual/en/class.deprecated.php

**Comment by macropay-solutions** [comment]
Starting from php 8.2 [AllowDynamicProperties](https://php.watch/versions/8.2/dynamic-properties-deprecated#AllowDynamicProperties) is another php attribute that might fall into this scenario.

**Comment by macropay-solutions** [comment]
SensitiveParameter https://www.php.net/manual/en/class.sensitiveparameter.php

**Comment by macropay-solutions** [comment]
ReturnTypeWillChange https://www.php.net/manual/en/class.returntypewillchange.php

**Comment by macropay-solutions** [comment]
We made a wiki about this https://github.com/macropay-solutions/maravel-framework/wiki/PHP-Native-Attributes
